### DISCUSSION TITLE
Error: null input file! in nvbootctrl

### DISCUSSION THREAD

**Discussion by albertbu** [OE4T/meta-tegra#1634] [General]
HI, I'm trying to implement the AB boot on our custom board, based on Orin NX. I succeed to build the redundant partitions, but when I try to check the status of the boot slots I get 

root@mach-jetson:ü# nvbootctrl verify
Info: variable BootChainFwStatus is not found.
Error: null input file!
Warning: Failed to get chip id from config

In my conf file I have 

USE_REDUNDANT_FLASH_LAYOUT_DEFAULT = "1"
IMAGE_INSTALL:append = " tegra-uefi-capsules  tegra-redundant-boot"
TNSPEC_BOOTDEV = "nvme0n1p1"

What am I doing wrong? 
Thanks in advance

**Comment by dwalkes** [comment]
Which branch?  Can you reproduce with https://github.com/OE4T/tegra-demo-distro and an Orin NX based MACHINE there with similar settings?

Based on the similar failures at https://github.com/OE4T/meta-tegra/issues/1531 my suspicion is that it has something to do with the MACHINE TEGRA_BUPGEN_SPECS but that's just a guess.  I would try to compare my setup with a working similar tegra-demo-distro setup as a starting point.

**Reply by albertbu** [reply]
I'm using scartgath - latest tegra version 
Is there a step-by-step guide to achieve this? 
I am not touching TEGRA_BUPGEN_SPECS , it should come from the orin-nano.inc file

**Reply by dwalkes** [reply]
> Is there a step-by-step guide to achieve this?

Do you mean to build tegra demo distro with scarthgap?  If so it's at https://github.com/OE4T/tegra-demo-distro/tree/scarthgap?tab=readme-ov-file#setting-up

**Reply by albertbu** [reply]
No, about implementing AB boot correctly

**Reply by dwalkes** [reply]
What you've got should be sufficient.  I suspect it will work on tegra-demo-distro on the scarthgap branch, I've used it there recently.

**Comment by albertbu** [comment]
HI 
I tried just now to pull the latest tegra. Same result. 
I'm on 
commit caf164e4917dc56d59990c72f9d8284abb6621f5 (HEAD, origin/scarthgap)
What can be the cause?

**Comment by albertbu** [comment]
Hi, 
where is the source code of nvbootctrl? Maybe I'll get more insights reading the code

**Reply by madisongh** [reply]
That tool is provided by NVIDIA in binary form only, unfortunately.

**Comment by albertbu** [comment]
Hi,
At least can I know which files is trying to access? I built a terga demo distro and of course it looks fine, but I’d like to know the exact mechanism and files needed by all the AB switching process
Thanks!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by dwalkes** [reply]
You can use strace for this.

**Comment by albertbu** [comment]
OK I found the root cause. Our build uses openrc instead of systemd, and some stuff is not in place

**Reply by dwalkes** [reply]
I just noticed this myself today, it can happen when `/etc/nv_boot_control.conf` is not present when `nvbootctrl` is invoked, which can happen when `nvbootctrl` is run before `setup-nv-boot-control.service` runs on systemd or the equivalent.

**Comment by dwalkes** [comment]
See related fix in https://github.com/OE4T/tegra-demo-distro/pull/371 and https://github.com/OE4T/tegra-demo-distro/pull/371/commits/6958260d276c9cb6d388fb190bf07c3148caabda for anyone noticing this on the swupdate demo implementation or a similar implementation.
